java/jdbcdriver-virtual-threads/src/main/java/com/oracle/dev/jdbc/PipelineVectorDemo.java

```diff
@@ -0,0 +1,789 @@
+      }
+    }
+  }
+
+  /**
+   * Loads the example table with facts about animals. This method uses
+   * the structured concurrency API (preview feature in JDK 22). Calls to
+   * {@link StructuredTaskScope#fork(Callable)} execute concurrent batch INSERTs
+   * into Oracle Database. The batch INSERTs are pipelined: This allows for
+   * concurrent progress of MANY statements from ONE JDBC connection.
+   */
+  static void loadTable(Connection connection)
+    throws IOException, InterruptedException, ExecutionException {
+
+    long start = System.currentTimeMillis();
+
+    // Stream text from a book about animals
+    String bookUrl =
+      "https://www.gutenberg.org/cache/epub/37959/pg37959.txt";
+    try (
+      Stream<String> paragraphStream = streamParagraphs(bookUrl);
+      ShutdownOnFailure taskScope = new ShutdownOnFailure()) {
+
+      // For every 100 paragraphs of text, execute a pipelined batch insert
+      paragraphStream.gather(Gatherers.windowFixed(100))
+        .forEach(paragraphs ->
+          taskScope.fork(() -> pipelineBatchInsert(connection, paragraphs)));
+
+      // Wait for all pipelined batch insert tasks to complete
+      taskScope.join();
+      taskScope.throwIfFailed();
+    }
+    finally {
+      long duration = System.currentTimeMillis() - start;
+      log(String.format("Table loaded in %.3f seconds", (duration / 1000f)));
+    }
+  }
+
+  /**
+   * Executes a pipelined batch INSERT using
+   * {@link OraclePreparedStatement#executeBatchAsyncOracle()}. The async method
+   * returns a Reactive Streams publisher, but this method blocks until the
+   * Publisher completes and returns the result; Blocking is no problem with
+   * virtual threads, because virtual threads are cheap!
+   */
+  static List<Long> pipelineBatchInsert(
+    Connection connection, List<String> lines)
+    throws SQLException {
+    log("Inserting " + lines.size() + " rows...");
+
+    try (PreparedStatement insert = connection.prepareStatement(
+      "INSERT INTO example(text) VALUES(?)")) {
+
+      for (String line : lines) {
+        insert.setString(1, line);
+        insert.addBatch();
+      }
+
+      // Pipeline the batch INSERT with executeBatchAsyncOracle
+      Flow.Publisher<Long> publisher =
+        insert.unwrap(OraclePreparedStatement.class)
+          .executeBatchAsyncOracle();
+
+      // No need for async/reactive callbacks. Just block the virtual thread.
+      List<Long> updateCounts = new ArrayList<>();
+      block(publisher, updateCounts::add);
+      return updateCounts;
+    }
+    finally {
+      log("Done inserting " + lines.size() + " rows");
+    }
+  }
+
+  /**
+   * Record class that stores the id and text column values of rows from the
+   * example table
+   */
+  private record TextRecord(int id, String text) { }
+
+  /**
+   * Updates the example table with embeddings for it's text data. This method
+   * uses the structured concurrency API (preview feature in JDK 22). Calls to
```

Comment:
It's preview feature of JDK21, isn't?

Reply:
We need JDK22 for the "Gatherers".

Reply:
Corrected in latest push. Elaborated that structured concurrency has been in preview since 21, and continues to be in the upcoming 23 release. 
Also, Gatherers are previewed in 22 and 23. I wanted to keep the focus on structured concurrency though. We could implement this demo without gatherers if we wanted to.